### PR TITLE
Revise, expand and update the documentation on solvers/

### PR DESCRIPTION
**Pull request by martin-cs**
Start work updating and revising the description of solvers/

Paging @peterschrammel

### CODE REVIEW

File: src/solvers/README.md [hunk 1]
```diff
@@ -6,59 +6,155 @@
...
+in CNF) over Boolean variables.  An SMT solver is a tool that
+implements decision procedures for the Satisfiability Modulo Theories
+class of problems.  CPROVER includes its own SMT solver, built on top
+of a SAT solver but can also interface to external solvers.
+
+CBMC and JBMC create formulae which describe some of the execution of parts of a
+program (see \ref goto-symex for how this is done) and then use a solver to see
+if there are any executions which break an assertion.  If the formula describing
+the execution and the formula describing the assertion are satisfiable then
+then it is possible for the assertion to fail and the assignment of the
```
Comment (peterschrammel) [inline src/solvers/README.md:27]: then then

File: src/solvers/README.md [hunk 2]
```diff
@@ -6,59 +6,155 @@
...
+variables can be used to build an error trace (see \ref goto-trace).  Thus
```
Comment (peterschrammel) [inline src/solvers/README.md]: I don't think this reference is accurate: \ref goto_tracet ?

File: src/solvers/README.md [hunk 3]
```diff
@@ -6,59 +6,155 @@
...
+Tools (and components) that implement decision procedures are often
+called solvers.  For example a SAT solver is a tool that implements a
+decision procedure for checking the satisfiability of formulae (often
+in CNF) over Boolean variables.  An SMT solver is a tool that
+implements decision procedures for the Satisfiability Modulo Theories
+class of problems.  CPROVER includes its own SMT solver, built on top
+of a SAT solver but can also interface to external solvers.
+
+CBMC and JBMC create formulae which describe some of the execution of parts of a
+program (see \ref goto-symex for how this is done) and then use a solver to see
```
Comment (peterschrammel) [inline src/solvers/README.md:24]: Is this ref accurate?

File: src/solvers/README.md [hunk 4]
```diff
@@ -6,59 +6,155 @@
...
+CBMC and JBMC.  Other tools make use of solvers in other ways to handle other
+problems.  It is important to distinguish between goto-models,
+goto-programs, etc. which describe programs and have a semantics in
+terms of execution and formula that have a semantics in terms of
+logic.  Solvers work with formulae and so have no notion of execution
+order, assignment, "before", branching, loops, exceptions,
+side-effects, function calls, etc.  All of these have to be described
+in the formula that are input if you want to reason about them.
+
+Other tools us solvers in different ways but the basic interface and
```
Comment (peterschrammel) [inline src/solvers/README.md]: use

File: src/solvers/README.md [hunk 5]
```diff
@@ -6,59 +6,155 @@
...
+software interface to a SAT solver : this is `propt`.  It uses the
+same `literalt` to refer to Boolean variables, just as `prop_convt`
+uses them to refer to logical expressions.  `land`, `lor`, `lxor` and
+so on allow gates to be constructed to express the formulae to be
+solved.  If `cnf_handled_well` is true then you may also use `lcnf` to
+build formulae.  Finally, `prop_solve` will run the decision procedure.
+
+As previously mentioned, many decision procedures reduce formulae to
+CNF and solve with a SAT solver.  `prop_conv_solvert` contains the
+foundations of this conversion.  It implements the `prop_conv` by
```
Comment (peterschrammel) [inline src/solvers/README.md]: prop_convt

File: src/solvers/README.md [hunk 6]
```diff
@@ -6,59 +6,155 @@
...
+having an instance of `propt` (a SAT solver) and reducing the
+expressions that are input into CNF.  The key entry point to this
+procedure is `convert` which then splits into `convert_boolean` (which
```
Comment (peterschrammel) [inline src/solvers/README.md]: prefix function names by the class name to make doxygen link it correctly

File: src/solvers/README.md (hunk 1 above)
Comment (martin-cs) [reply src/solvers/README.md:27]: Fixed.

File: src/solvers/README.md (hunk 2 above)
Comment (martin-cs) [reply src/solvers/README.md]: Fixed.

File: src/solvers/README.md (hunk 3 above)
Comment (martin-cs) [reply src/solvers/README.md:24]: Yes in the general sense; it's in that area of the code.

File: src/solvers/README.md (hunk 4 above)
Comment (martin-cs) [reply src/solvers/README.md]: Done.

File: src/solvers/README.md (hunk 5 above)
Comment (martin-cs) [reply src/solvers/README.md]: Done.

File: src/solvers/README.md (hunk 6 above)
Comment (martin-cs) [reply src/solvers/README.md]: Done; I think.

**Comment by martin-cs** [comment]
Paging @peterschrammel and @pkesseli : I have fixed the comments and rebased.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dda311ead11fc9d44b56967873b436be005b860d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99034406

File: src/solvers/README.md [hunk 7]
```diff
@@ -6,47 +6,156 @@
...
-  `convert` which takes an `exprt` and converts it to the appropriate,
-  solver specific, data structures and `dec_solve` (inherited from
-  `decision_proceduret`) which invokes the actual decision procedures.
-  Individual decision procedures (named `*_dect`) objects can be created
-  but `prop_convt` is the preferred interface for code that uses them.
-
-* flattening/:   A library that converts operations to bit-vectors,
+This directory contains most of the decision procedure code in CPROVER.
+A decision procedure is an algorithm which can check if a set of logical
+statements is satisfiable, i.e. if there is an value for each variable which
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
statements is satisfiable, i.e. if there is a value for each variable which
```

File: src/solvers/README.md [hunk 8]
```diff
@@ -6,47 +6,156 @@
...
+variables can be used to build an error trace (see \ref goto_tracet).  Thus
+the performance and capability of the solver used is crucial to the utility of
+CBMC and JBMC.  Other tools make use of solvers in other ways to handle other
+problems.  It is important to distinguish between goto-models,
+goto-programs, etc. which describe programs and have a semantics in
+terms of execution and formula that have a semantics in terms of
+logic.  Solvers work with formulae and so have no notion of execution
+order, assignment, "before", branching, loops, exceptions,
+side-effects, function calls, etc.  All of these have to be described
+in the formula that are input if you want to reason about them.
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
in the formula presented to the decision procedure if you want to reason about them.
```

File: src/solvers/README.md [hunk 9]
```diff
@@ -6,47 +6,156 @@
...
+
+Other tools use solvers in different ways but the basic interface and
+ideas remain the same.
+
+
+\section solvers-interfaces Key Interfaces
+
+The most basic interface is `decision_proceduret` .  It gives the
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
The most basic interface is `decision_proceduret`.  It gives the
```

File: src/solvers/README.md [hunk 10]
```diff
@@ -6,47 +6,156 @@
...
+
+Within decision procedures it is common to reduce the logical
+expressions to equivalent expressions in a simpler language.  This is
+similar to what a compiler will do in reducing higher-level language
+constructs to simple, assembler like instructions.  This, of course,
+relies on having a decision procedure for the simpler language, just
+as a compiler relies on you having a processor that can handle the
+assembler.  One of the popular choices of "processor" for decision
+procedures are SAT solvers.  These handle a very restricted language
+in which all variables are simple Booleans and all formula are just
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
in which all variables are simple Booleans and all formulae are just
```

File: src/solvers/README.md [hunk 11]
```diff
@@ -6,47 +6,156 @@
...
+made of logical gates.  By keeping their input simple, they can be
+made very fast and efficient; kind of like RISC processors.  Like
+processors, creating a good SAT solver is a very specialised skill, so
+CPROVER uses third-party SAT solvers.  By default this is MiniSAT, but
+other are supported (see the `sat/` directory).  To do this it needs a
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
others are supported (see the `sat/` directory).  To do this it needs a
```

File: src/solvers/README.md [hunk 12]
```diff
@@ -6,47 +6,156 @@
...
+  `cnft` and `cnf_solvert` which give default implements `propt`s gate
+  functions (`land`, `lxor`, etc.) in terms of `lcnf` as most modern
+  SAT solvers only have interfaces for handling CNF, not logical
+  gates.  The various satcheck_* files implement the `propt`
+  interfaces (generally with the `cnf_solvert` additions /
+  simplifications) by connecting to various third-party SAT solvers.
+  Finally `satcheck.h` use the build time flags to pick which SAT
+  solvers are available and which should be used as default.
+
+* `qbf/`:   An equivalent of `sat/' for QBF solvers.  These extend
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
* `qbf/`:   An equivalent of `sat/` for QBF solvers.  These extend
```

File: src/solvers/README.md [hunk 13]
```diff
@@ -6,47 +6,156 @@
...
+  SMT-LIB 2. With the `–fpa` option, this output mode will not flatten
+  the floating point arithmetic and instead output SMT-LIB floating point standard.
+
+* `floatbv/`:    This library contains the code that is used to
+  convert floating point variables (`floatbv`) to bit vectors
+  (`bv`). This is referred to as ‘bit-blasting’ and is called in the
+  `solver` code during conversion to SAT or SMT. It also contains the
+  abstraction code described in the FMCAD09 paper.
+
+*. `lowering/`:  These are `exprt` to `exprt` reductions of operations
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
* `lowering/`:  These are `exprt` to `exprt` reductions of operations
```

File: src/solvers/README.md [hunk 14]
```diff
@@ -6,47 +6,156 @@
...
+   rather than the `exprt` to `bvt` reductions in `flattening/`
+   allowing them to be used in SMT solvers which do not inherit from
+   `prop_conv_solvert`.
+
+*. `refinement/`: Solvers that build on the `bv_pointerst` solver
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
* `refinement/`: Solvers that build on the `bv_pointerst` solver
```

File: src/solvers/README.md [hunk 15]
```diff
@@ -6,47 +6,156 @@
...
+   interface and specialise the handling of certain operations to
+   improve performance.
+
+*. `miniBDD/`:  A canonical representation of Boolean formulae.
```
Comment (tautschnig) [inline src/solvers/README.md]:
```suggestion
* `miniBDD/`:  A canonical representation of Boolean formulae.
```

**Review by tautschnig — APPROVED** [review]
A couple of minor tweaks.

**Comment by martin-cs** [comment]
@tautschnig : thanks for the fixes; I've applied all of them via the github interface (I think).  Feel free to merge while I'm away.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 71592f806758ea059aa6d084b4e7779656b230e6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99314843

**Comment by martin-cs** [comment]
Squashed.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0773e235442dcc3ded1fc2b66b186aa1476eac7b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99327676

**Pull request merged**